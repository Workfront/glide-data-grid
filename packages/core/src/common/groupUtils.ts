import { GridRow, GridRowKind } from '../data-grid/data-grid-types';
import type { RowGroup } from '../data-editor/use-groups';

export const getNestedGroupRowsCount = (group: RowGroup): number => {
  if (group === undefined) return 0;

  return (
    group.rowsCount +
    group.groups.reduce((acc, currentGroup) => acc + getNestedGroupRowsCount(currentGroup), 0)
  );
};

export const recursiveFlattenGroups = (
  groups: readonly RowGroup[],
  rows: GridRow[],
  level = 1,
  rowIndex = 0,
  hasTrailingRow: boolean
): { rows: GridRow[]; rowIndex: number } => {
  for (const group of groups) {
    rows.push({
      kind: GridRowKind.Group,
      level: level,
      name: group.name,
      expanded: group.expanded,
      id: group.id,
      allowOverlay: false,
    });

    if (!group.expanded) {
      rowIndex += getNestedGroupRowsCount(group);
      continue;
    }

    if (group.groups.length > 0) {
      rowIndex = recursiveFlattenGroups(
        group.groups,
        rows,
        level + 1,
        rowIndex,
        hasTrailingRow
      ).rowIndex;
    } else {
      for (let i = 0; i < group.rowsCount; i++) {
        rows.push({
          kind: GridRowKind.GroupContent,
          level: level + 1,
          index: rowIndex + i,
          groupId: group.id,
        });
      }
      rowIndex += group.rowsCount;

      if (hasTrailingRow) {
        rows.push({
          kind: GridRowKind.NewRow,
          index: rowIndex,
          level: level + 1,
          groupId: group.id,
        });
      }
    }
  }

  return { rows, rowIndex };
};

export const flattenGroups = (groups: readonly RowGroup[], hasTrailingRow: boolean): GridRow[] => {
  const rows: GridRow[] = [];
  return recursiveFlattenGroups(groups, rows, 1, 0, hasTrailingRow).rows;
};

export const findGroupById = (currentGroups: RowGroup[], groupId: string): RowGroup | undefined => {
  for (const group of currentGroups) {
    if (group.id === groupId) {
      return group;
    }

    const foundGroup = findGroupById(group.groups, groupId);
    if (foundGroup) {
      return foundGroup;
    }
  }
};

export const deleteGroupByGroupRowId = (
  rootGroups: RowGroup[],
  targetGroupRowId: string
): boolean => {
  const groupToDelete = findGroupById(rootGroups, targetGroupRowId);

  const parentId = groupToDelete?.parentId ?? '';

  const parentGroup =
    parentId !== undefined && parentId !== ''
      ? findGroupById(rootGroups, parentId)?.groups
      : rootGroups;

  if (!parentGroup || !groupToDelete) {
    return false;
  }

  // Find the index of the group to delete within its parent's groups
  const index = parentGroup.indexOf(groupToDelete);

  if (index >= 0) {
    parentGroup.splice(index, 1);

    // Recursively delete empty parent groups
    if (parentGroup.length === 0) {
      deleteGroupByGroupRowId(rootGroups, parentId);
    }

    return true;
  }

  return false;
};
