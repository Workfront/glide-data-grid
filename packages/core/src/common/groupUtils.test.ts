import type { RowGroup } from "../data-editor/use-groups";
import { deleteGroupByGroupRowId, flattenGroups } from "./groupUtils";

const getGroups = (): RowGroup[] => {
  return [
    {
      name: 'Group 1',
      rowsCount: 0,
      expanded: true,
      id: '1',
      parentId: '',
      groups: [
        {
          name: 'Group 1-1',
          rowsCount: 0,
          id: '1-1',
          parentId: '1',
          groups: [
            {
              name: 'Group 1-1-1',
              rowsCount: 1,
              groups: [],
              expanded: true,
              id: '1-1-1',
              parentId: '1-1',
            },
            {
              name: 'Group 1-1-2',
              rowsCount: 1,
              groups: [],
              expanded: true,
              id: '1-1-2',
              parentId: '1-1',
            },
          ],
          expanded: true,
        },
        { name: 'Group 1-2', rowsCount: 1, groups: [], expanded: true, id: '1-2', parentId: '1' },
      ],
    },
    {
      name: 'Group 2',
      rowsCount: 0,
      expanded: true,
      id: '2',
      parentId: '',
      groups: [
        {
          name: 'Group 2-1',
          rowsCount: 0,
          id: '2-1',
          parentId: '2',
          groups: [
            {
              name: 'Group 2-1-1',
              rowsCount: 1,
              groups: [],
              expanded: true,
              id: '2-1-1',
              parentId: '2-1',
            },
            {
              name: 'Group 2-1-2',
              rowsCount: 1,
              groups: [],
              expanded: true,
              id: '2-1-2',
              parentId: '2-1',
            },
          ],
          expanded: true,
        },
        { name: 'Group 2-2', rowsCount: 1, groups: [], expanded: true, id: '2-2', parentId: '2' },
      ],
    },
    {
      name: 'Group 3',
      rowsCount: 0,
      expanded: true,
      id: '3',
      parentId: '',
      groups: [
        {
          name: 'Group 3-1',
          rowsCount: 0,
          id: '3-1',
          parentId: '3',
          groups: [
            {
              name: 'Group 3-1-1',
              rowsCount: 1,
              groups: [],
              expanded: true,
              id: '3-1-1',
              parentId: '3-1',
            },
            {
              name: 'Group 3-1-2',
              rowsCount: 1,
              groups: [],
              expanded: true,
              id: '3-1-2',
              parentId: '3-1',
            },
          ],
          expanded: true,
        },
        { name: 'Group 3-2', rowsCount: 1, groups: [], expanded: true, id: '3-2', parentId: '3-1' },
      ],
    },
  ];
};

describe("flattenGroups", () => {

    it("matches snapshot", () => {
        const groups = getGroups();
        expect(flattenGroups(groups, true)).toMatchSnapshot();
    });

    it("collapses first group", () => {
        const groups = getGroups();
        groups[0].expanded = false;
        expect(flattenGroups(groups, true)).toMatchSnapshot();
    });

    it("has single level group", () => {
        const groups = [
            {
                name: `Group 1`,
                rowsCount: 3,
                expanded: true,
                id: "1",
        groups: [],
        parentId: '',
      },
      {
        name: `Group 2`,
        rowsCount: 3,
        expanded: true,
        id: "2",
        groups: [],
        parentId: '',
      },
      {
        name: `Group 2`,
        rowsCount: 3,
        expanded: true,
        id: "3",
        groups: [],
        parentId: '',
            },
        ];

        groups[1].expanded = false;
        expect(flattenGroups(groups, true)).toMatchSnapshot();
    });

  it('should delete a group by groupRowId and return true', () => {
    const targetGroupRowId = '1-1-1';
    const mockData = getGroups();
    deleteGroupByGroupRowId(mockData, targetGroupRowId);

    // Ensure the group is deleted
    expect(mockData).toMatchSnapshot();
  });
  it('should return false when the groupRowId does not exist', () => {
    const targetGroupRowId = '4-3-1';
    const mockData = getGroups();
    const result = deleteGroupByGroupRowId(mockData, targetGroupRowId);

    expect(result).toBe(false);
  });

  it('should handle deleting a root-level group', () => {
    const targetGroupRowId = '1-1-1';
    const mockData = getGroups();
    const result = deleteGroupByGroupRowId(mockData, targetGroupRowId);

    expect(result).toBe(true);
    // Ensure the root-level group is deleted
    expect(mockData).toMatchSnapshot();
  });
});
