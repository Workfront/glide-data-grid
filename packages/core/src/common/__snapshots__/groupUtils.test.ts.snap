// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flattenGroups collapses first group 1`] = `
Array [
  Object {
    "allowOverlay": false,
    "expanded": false,
    "id": "1",
    "kind": "group",
    "level": 1,
    "name": "Group 1",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2",
    "kind": "group",
    "level": 1,
    "name": "Group 2",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-1",
    "kind": "group",
    "level": 2,
    "name": "Group 2-1",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-1-1",
    "kind": "group",
    "level": 3,
    "name": "Group 2-1-1",
  },
  Object {
    "groupId": "2-1-1",
    "index": 3,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "2-1-1",
    "index": 4,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-1-2",
    "kind": "group",
    "level": 3,
    "name": "Group 2-1-2",
  },
  Object {
    "groupId": "2-1-2",
    "index": 4,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "2-1-2",
    "index": 5,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-2",
    "kind": "group",
    "level": 2,
    "name": "Group 2-2",
  },
  Object {
    "groupId": "2-2",
    "index": 5,
    "kind": "group-content",
    "level": 3,
  },
  Object {
    "groupId": "2-2",
    "index": 6,
    "kind": "new-row",
    "level": 3,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3",
    "kind": "group",
    "level": 1,
    "name": "Group 3",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-1",
    "kind": "group",
    "level": 2,
    "name": "Group 3-1",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-1-1",
    "kind": "group",
    "level": 3,
    "name": "Group 3-1-1",
  },
  Object {
    "groupId": "3-1-1",
    "index": 6,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "3-1-1",
    "index": 7,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-1-2",
    "kind": "group",
    "level": 3,
    "name": "Group 3-1-2",
  },
  Object {
    "groupId": "3-1-2",
    "index": 7,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "3-1-2",
    "index": 8,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-2",
    "kind": "group",
    "level": 2,
    "name": "Group 3-2",
  },
  Object {
    "groupId": "3-2",
    "index": 8,
    "kind": "group-content",
    "level": 3,
  },
  Object {
    "groupId": "3-2",
    "index": 9,
    "kind": "new-row",
    "level": 3,
  },
]
`;

exports[`flattenGroups has single level group 1`] = `
Array [
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "1",
    "kind": "group",
    "level": 1,
    "name": "Group 1",
  },
  Object {
    "groupId": "1",
    "index": 0,
    "kind": "group-content",
    "level": 2,
  },
  Object {
    "groupId": "1",
    "index": 1,
    "kind": "group-content",
    "level": 2,
  },
  Object {
    "groupId": "1",
    "index": 2,
    "kind": "group-content",
    "level": 2,
  },
  Object {
    "groupId": "1",
    "index": 3,
    "kind": "new-row",
    "level": 2,
  },
  Object {
    "allowOverlay": false,
    "expanded": false,
    "id": "2",
    "kind": "group",
    "level": 1,
    "name": "Group 2",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3",
    "kind": "group",
    "level": 1,
    "name": "Group 2",
  },
  Object {
    "groupId": "3",
    "index": 6,
    "kind": "group-content",
    "level": 2,
  },
  Object {
    "groupId": "3",
    "index": 7,
    "kind": "group-content",
    "level": 2,
  },
  Object {
    "groupId": "3",
    "index": 8,
    "kind": "group-content",
    "level": 2,
  },
  Object {
    "groupId": "3",
    "index": 9,
    "kind": "new-row",
    "level": 2,
  },
]
`;

exports[`flattenGroups matches snapshot 1`] = `
Array [
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "1",
    "kind": "group",
    "level": 1,
    "name": "Group 1",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "1-1",
    "kind": "group",
    "level": 2,
    "name": "Group 1-1",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "1-1-1",
    "kind": "group",
    "level": 3,
    "name": "Group 1-1-1",
  },
  Object {
    "groupId": "1-1-1",
    "index": 0,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "1-1-1",
    "index": 1,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "1-1-2",
    "kind": "group",
    "level": 3,
    "name": "Group 1-1-2",
  },
  Object {
    "groupId": "1-1-2",
    "index": 1,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "1-1-2",
    "index": 2,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "1-2",
    "kind": "group",
    "level": 2,
    "name": "Group 1-2",
  },
  Object {
    "groupId": "1-2",
    "index": 2,
    "kind": "group-content",
    "level": 3,
  },
  Object {
    "groupId": "1-2",
    "index": 3,
    "kind": "new-row",
    "level": 3,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2",
    "kind": "group",
    "level": 1,
    "name": "Group 2",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-1",
    "kind": "group",
    "level": 2,
    "name": "Group 2-1",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-1-1",
    "kind": "group",
    "level": 3,
    "name": "Group 2-1-1",
  },
  Object {
    "groupId": "2-1-1",
    "index": 3,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "2-1-1",
    "index": 4,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-1-2",
    "kind": "group",
    "level": 3,
    "name": "Group 2-1-2",
  },
  Object {
    "groupId": "2-1-2",
    "index": 4,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "2-1-2",
    "index": 5,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "2-2",
    "kind": "group",
    "level": 2,
    "name": "Group 2-2",
  },
  Object {
    "groupId": "2-2",
    "index": 5,
    "kind": "group-content",
    "level": 3,
  },
  Object {
    "groupId": "2-2",
    "index": 6,
    "kind": "new-row",
    "level": 3,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3",
    "kind": "group",
    "level": 1,
    "name": "Group 3",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-1",
    "kind": "group",
    "level": 2,
    "name": "Group 3-1",
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-1-1",
    "kind": "group",
    "level": 3,
    "name": "Group 3-1-1",
  },
  Object {
    "groupId": "3-1-1",
    "index": 6,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "3-1-1",
    "index": 7,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-1-2",
    "kind": "group",
    "level": 3,
    "name": "Group 3-1-2",
  },
  Object {
    "groupId": "3-1-2",
    "index": 7,
    "kind": "group-content",
    "level": 4,
  },
  Object {
    "groupId": "3-1-2",
    "index": 8,
    "kind": "new-row",
    "level": 4,
  },
  Object {
    "allowOverlay": false,
    "expanded": true,
    "id": "3-2",
    "kind": "group",
    "level": 2,
    "name": "Group 3-2",
  },
  Object {
    "groupId": "3-2",
    "index": 8,
    "kind": "group-content",
    "level": 3,
  },
  Object {
    "groupId": "3-2",
    "index": 9,
    "kind": "new-row",
    "level": 3,
  },
]
`;

exports[`flattenGroups should delete a group by groupRowId and return true 1`] = `
Array [
  Object {
    "expanded": true,
    "groups": Array [
      Object {
        "expanded": true,
        "groups": Array [
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "1-1-2",
            "name": "Group 1-1-2",
            "parentId": "1-1",
            "rowsCount": 1,
          },
        ],
        "id": "1-1",
        "name": "Group 1-1",
        "parentId": "1",
        "rowsCount": 0,
      },
      Object {
        "expanded": true,
        "groups": Array [],
        "id": "1-2",
        "name": "Group 1-2",
        "parentId": "1",
        "rowsCount": 1,
      },
    ],
    "id": "1",
    "name": "Group 1",
    "parentId": "",
    "rowsCount": 0,
  },
  Object {
    "expanded": true,
    "groups": Array [
      Object {
        "expanded": true,
        "groups": Array [
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "2-1-1",
            "name": "Group 2-1-1",
            "parentId": "2-1",
            "rowsCount": 1,
          },
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "2-1-2",
            "name": "Group 2-1-2",
            "parentId": "2-1",
            "rowsCount": 1,
          },
        ],
        "id": "2-1",
        "name": "Group 2-1",
        "parentId": "2",
        "rowsCount": 0,
      },
      Object {
        "expanded": true,
        "groups": Array [],
        "id": "2-2",
        "name": "Group 2-2",
        "parentId": "2",
        "rowsCount": 1,
      },
    ],
    "id": "2",
    "name": "Group 2",
    "parentId": "",
    "rowsCount": 0,
  },
  Object {
    "expanded": true,
    "groups": Array [
      Object {
        "expanded": true,
        "groups": Array [
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "3-1-1",
            "name": "Group 3-1-1",
            "parentId": "3-1",
            "rowsCount": 1,
          },
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "3-1-2",
            "name": "Group 3-1-2",
            "parentId": "3-1",
            "rowsCount": 1,
          },
        ],
        "id": "3-1",
        "name": "Group 3-1",
        "parentId": "3",
        "rowsCount": 0,
      },
      Object {
        "expanded": true,
        "groups": Array [],
        "id": "3-2",
        "name": "Group 3-2",
        "parentId": "3-1",
        "rowsCount": 1,
      },
    ],
    "id": "3",
    "name": "Group 3",
    "parentId": "",
    "rowsCount": 0,
  },
]
`;

exports[`flattenGroups should handle deleting a root-level group 1`] = `
Array [
  Object {
    "expanded": true,
    "groups": Array [
      Object {
        "expanded": true,
        "groups": Array [
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "1-1-2",
            "name": "Group 1-1-2",
            "parentId": "1-1",
            "rowsCount": 1,
          },
        ],
        "id": "1-1",
        "name": "Group 1-1",
        "parentId": "1",
        "rowsCount": 0,
      },
      Object {
        "expanded": true,
        "groups": Array [],
        "id": "1-2",
        "name": "Group 1-2",
        "parentId": "1",
        "rowsCount": 1,
      },
    ],
    "id": "1",
    "name": "Group 1",
    "parentId": "",
    "rowsCount": 0,
  },
  Object {
    "expanded": true,
    "groups": Array [
      Object {
        "expanded": true,
        "groups": Array [
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "2-1-1",
            "name": "Group 2-1-1",
            "parentId": "2-1",
            "rowsCount": 1,
          },
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "2-1-2",
            "name": "Group 2-1-2",
            "parentId": "2-1",
            "rowsCount": 1,
          },
        ],
        "id": "2-1",
        "name": "Group 2-1",
        "parentId": "2",
        "rowsCount": 0,
      },
      Object {
        "expanded": true,
        "groups": Array [],
        "id": "2-2",
        "name": "Group 2-2",
        "parentId": "2",
        "rowsCount": 1,
      },
    ],
    "id": "2",
    "name": "Group 2",
    "parentId": "",
    "rowsCount": 0,
  },
  Object {
    "expanded": true,
    "groups": Array [
      Object {
        "expanded": true,
        "groups": Array [
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "3-1-1",
            "name": "Group 3-1-1",
            "parentId": "3-1",
            "rowsCount": 1,
          },
          Object {
            "expanded": true,
            "groups": Array [],
            "id": "3-1-2",
            "name": "Group 3-1-2",
            "parentId": "3-1",
            "rowsCount": 1,
          },
        ],
        "id": "3-1",
        "name": "Group 3-1",
        "parentId": "3",
        "rowsCount": 0,
      },
      Object {
        "expanded": true,
        "groups": Array [],
        "id": "3-2",
        "name": "Group 3-2",
        "parentId": "3-1",
        "rowsCount": 1,
      },
    ],
    "id": "3",
    "name": "Group 3",
    "parentId": "",
    "rowsCount": 0,
  },
]
`;
